File: 61_RotateList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null || head.next == null) {
            return head;
        }


        int len = 1;
        ListNode tail = head;
        while (tail.next != null) {
            len += 1;
            tail = tail.next;
        }
        tail.next = head;

        k %= len;
        int skip = len - k;
        ListNode cur = head;
        for (int i = 0; i < skip - 1; i++) {
            cur = cur.next;
        }
        head = cur.next;
        cur.next = null;
        return head;
    }
}

File: c++/61_RotateList.cpp
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if (!head) return head;
        int n = 0;
        auto tail = head;
        for (auto p = head; p; p = p->next) {
            tail = p;
            n++;
        }
        k %= n;
        if (!k) return head;
        auto p = head;
        for (int i = 0; i < n - k - 1; i++) p = p->next;
        tail->next = head;
        head = p->next;
        p->next = nullptr;
        return head;
    }
};